resolvers/webpack/index.js

```diff
@@ -79,75 +82,137 @@ exports.resolve = function (source, file, settings) {
+    var webpackResolveOpts = { basedir: path.dirname(webpackFilename) }
+
+    webpackRequire = function (id) {
+      return require(resolve.sync(id, webpackResolveOpts))
+    }
+  } catch (e) {
+    // Something has gone wrong (or we're in a test). Use our own bundled
+    // enhanced-resolve.
+    webpackRequire = require
+  }
+
+  var enhancedResolvePackage = webpackRequire('enhanced-resolve/package.json')
+  var enhancedResolveVersion = enhancedResolvePackage.version
+  log('enhanced-resolve version:', enhancedResolveVersion)
+
+  var resolveConfig = webpackConfig.resolve || {}
+
+  if (semver.major(enhancedResolveVersion) === 2) {
+    return createWebpack2ResolveSync(webpackRequire, resolveConfig)
+  }
+
+  return createWebpack1ResolveSync(webpackRequire, resolveConfig, webpackConfig.plugins)
+}
+
+function createWebpack2ResolveSync(webpackRequire, resolveConfig) {
+  var EnhancedResolve = webpackRequire('enhanced-resolve')
 
-var SyncNodeJsInputFileSystem = require('enhanced-resolve/lib/SyncNodeJsInputFileSystem')
-var syncFS = new SyncNodeJsInputFileSystem()
+  return EnhancedResolve.create.sync(assign({}, resolveConfig, webpack2DefaultResolveConfig))
+}
 
-var ModuleAliasPlugin = require('enhanced-resolve/lib/ModuleAliasPlugin')
-var ModulesInDirectoriesPlugin = require('enhanced-resolve/lib/ModulesInDirectoriesPlugin')
-var ModulesInRootPlugin = require('enhanced-resolve/lib/ModulesInRootPlugin')
-var ModuleAsFilePlugin = require('enhanced-resolve/lib/ModuleAsFilePlugin')
-var ModuleAsDirectoryPlugin = require('enhanced-resolve/lib/ModuleAsDirectoryPlugin')
-var DirectoryDescriptionFilePlugin = require('enhanced-resolve/lib/DirectoryDescriptionFilePlugin')
-var DirectoryDefaultFilePlugin = require('enhanced-resolve/lib/DirectoryDefaultFilePlugin')
-var FileAppendPlugin = require('enhanced-resolve/lib/FileAppendPlugin')
-var ResultSymlinkPlugin = require('enhanced-resolve/lib/ResultSymlinkPlugin')
-var DirectoryDescriptionFileFieldAliasPlugin =
-  require('enhanced-resolve/lib/DirectoryDescriptionFileFieldAliasPlugin')
+/**
+ * webpack 2 defaults:
+ * https://github.com/webpack/webpack/blob/v2.1.0-beta.20/lib/WebpackOptionsDefaulter.js#L72-L87
+ * @type {Object}
+ */
+var webpack2DefaultResolveConfig = {
+  unsafeCache: true, // Probably a no-op, since how can we cache anything at all here?
+  modules: ['node_modules'],
+  extensions: ['.js', '.json'],
+  aliasFields: ['browser'],
+  mainFields: ['browser', 'module', 'main'],
+}
 
 // adapted from tests &
 // https://github.com/webpack/webpack/blob/v1.13.0/lib/WebpackOptionsApply.js#L322
-function createResolver(resolve, otherPlugins) {
-  var resolver = new Resolver(syncFS)
+function createWebpack1ResolveSync(webpackRequire, resolveConfig, plugins) {
+  var Resolver = webpackRequire('enhanced-resolve/lib/Resolver')
+  var SyncNodeJsInputFileSystem = webpackRequire('enhanced-resolve/lib/SyncNodeJsInputFileSystem')
```

Comment:
https://github.com/webpack/enhanced-resolve/blob/v0.9.1/lib/SyncNodeJsInputFileSystem.js

This file system object has no caching attached to it AFAICT, so the penalty for constructing it on-the-fly should be trivial.